### PR TITLE
Added getting builds by definition name

### PR DESCRIPTION
**Pull request by awanlin**
Signed-off-by: Andre Wanlin <awanlin@rapidrtc.com>

## Hey, I just made a Pull Request!

Added getting builds by definition name which will lay the ground work for #7853 

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by awanlin** [comment]
@defenestration if you have some time to review this, that would be great! By getting Builds using the Build Definition Name we don't need to use Azure Repos. Once these changes get approved and merged in I'll work on adding this to the frontend

File: plugins/azure-devops-backend/api-report.md
Comment (benjdlambert) [inline plugins/azure-devops-backend/api-report.md:24]: @awanlin this looks like a breaking change here right?

**Comment by defenestration** [comment]
Hello,

Sorry it took me a while to notice this. 

I was able to spin up the fork/branch, and copying my app-settings.local.yaml and run it. After looking through the changes in this PR i see i can resolve some data at 
http://localhost:7000/api/azure-devops/definition-builds/Api%20Template/158 

Which corresponds to one of the "Pipelines" that is created in the AZDO project in which the url you can see the `definitionId`:
https://dev.azure.com/myOrg/API%20Template/_build?definitionId=158


``` json
[
  {
    "id": 125911,
    "title": "Deploy New API Service - Deploy YamlDeployment.b996656982257feb557e4a1f08085125cc48a0c2",
    "link": "https://dev.azure.com/myOrg/3cf4769a-973a-4c81-aeaf-0205982e4a7d/_build/results?buildId=125911",
    "status": 2,
    "result": 8,
    "queueTime": "2021-11-04T20:24:03.797Z",
    "startTime": "2021-11-08T14:13:03.349Z",
    "finishTime": "2021-11-08T14:13:40.014Z",
    "source": "refs/heads/YamlDeployment (b9966569)",
    "uniqueName": "a.user@myOrg.com"
  },
  {
    "id": 126175,
    "title": "Build New API Service - main.da2af1a1a1ba5dd26a67c37c71a9a6f806f81353",
    "link": "https://dev.azure.com/myOrg/3cf4769a-973a-4c81-aeaf-0205982e4a7d/_build/results?buildId=126175",
    "status": 2,
    "result": 2,
    "queueTime": "2021-11-08T00:00:06.592Z",
    "startTime": "2021-11-08T00:04:15.126Z",
    "finishTime": "2021-11-08T00:06:00.206Z",
    "source": "refs/heads/main (da2af1a1)",
    "uniqueName": "xxxxxxxx-0000-8888-8000-000000000000@2c895908-04e0-4952-89fd-54b0046d6288"
  },
  ... 
]
```

So assuming I did things correctly, it seems backstage at least has access to this data now, though I am not sure if we can pull it into the UI yet? If there is a way let me know how I could do that. I feel I lucked out getting this far!

**Comment by awanlin** [comment]
@defenestration, no worries at all. The idea is that this is the first set of changes to make it as easy as adding an annotation like this to your entity: `dev.azure.com/project-definition: <project-name>/<definition-name>`. Once this gets approved I'll work on adding the needed bit to the front end.

I was just looking to see if you could review the code and it seems that this is working for you so that's good :)

File: .changeset/rare-toes-burn.md
Comment (Rugvip) [inline .changeset/rare-toes-burn.md:2]: To @benjdlambert point, this would be a breaking change right? so `minor`

**Review by Rugvip — COMMENTED** [review]
👍 , but not diving too deep into review :grin:

Just wanna make sure changeset is good to go though

File: .changeset/rare-toes-burn.md
Comment (awanlin) [reply .changeset/rare-toes-burn.md:2]: Yes, but I'd like to refactor this, I think there is a non-break approach that I'd like to take that will also improve a few things

File: plugins/azure-devops-backend/api-report.md
Comment (awanlin) [reply plugins/azure-devops-backend/api-report.md:24]: Yes, sorry for the slow reply. I'd like to refactor this, I think there is a non-break approach that I'd like to take that will also improve a few things

**Comment by awanlin** [comment]
I've add a new API that we'll use to get builds that allows the ability to provide a repo name or a build definition name. It also is more extensible in the future if we need to add on to it. I've undone the changes that would have made this a breaking change.

This is ready to be reviewed whenever you have the time @benjdlambert @Rugvip

File: plugins/azure-devops-backend/api-report.md
Comment (Rugvip) [reply plugins/azure-devops-backend/api-report.md:24]: 👍 , it's still breaking because of the additions of required methods in the API, but the impact is ofc smaller. Best make it a `minor` change tho

**Review by Rugvip — COMMENTED** [review]
Looks good! 👍 

Just a `minor` note on the changeset

**Comment by freben** [comment]
@awanlin We ended up wanting to change this to `minor` right? Since the addition of methods will be breaking either way. Let us know what you think!

File: plugins/azure-devops-backend/api-report.md
Comment (awanlin) [reply plugins/azure-devops-backend/api-report.md:24]: I've updated this to `minor` now, thanks 👍

**Comment by awanlin** [comment]
@marleypowell when you have some time can you do a sanity check on this, please? Never hurts to have another set of eyes on this

File: plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts
Comment (marleypowell) [inline plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts:173]: Do you need to specify these optional parameters?
```suggestion
    return client.getDefinitions(projectName, definitionName);
```

File: .changeset/rare-toes-burn.md
Comment (marleypowell) [inline .changeset/rare-toes-burn.md:2]: Is there any breaking changes? If not them this is probably a `patch`.

File: plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts
Comment (marleypowell) [inline plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts:180]: Could the `definitions` parameter be added to the existing `getBuildList` method? It doesn't look like this method is doing much differently.

File: plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts
Comment (marleypowell) [inline plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts:231]: I'm assuming this is to cast from `number | undefined[]` to `number[]`? You can do this so it ensures there's no undefined values:
```suggestion
      definitions = buildDefinitions.map(bd => bd.id).filter((bd ): bd is number => Boolean(bd ));
```

File: plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts
Comment (marleypowell) [inline plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts:238]: Does this work?
```suggestion
    const buildRuns: BuildRun[] = builds.map(mappedBuildRun);
```

**Review by marleypowell — APPROVED** [review]
Looks good! A few suggestions added.

File: plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts
Comment (awanlin) [reply plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts]: This works, I remember having issues with the other API calls so put them in assuming I needed them here too, thanks for pointing this out 👍

File: plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts
Comment (awanlin) [reply plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts]: This is cool, I've updated it to use this

File: plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts
Comment (awanlin) [reply plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts]: Yes, that works and I've gone this route

File: plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts
Comment (awanlin) [reply plugins/azure-devops-backend/src/api/AzureDevOpsApi.ts:282]: I'm trying to avoid breaking changes

File: .changeset/rare-toes-burn.md
Comment (awanlin) [reply .changeset/rare-toes-burn.md:2]: Yeah, I don't think there are

**Comment by awanlin** [comment]
OK, I've finally got back into this and resolved the feedback. Where I'm a little lost is on the breaking changes part. There should not be any so I think changing this to `minor` makes sense. But I haven't built a ton of APIs so would like some help clearing this up, maybe I missed something. @Rugvip @benjdlambert @marleypowell looking for your advice, please 🤔

**Comment by jhaals** [comment]
Hi @awanlin, I took a peak at your PR and I don't see any breaking changes. New endpoints added and no change in existing behaviour. So in that case it should be a `minor` as the version of this plugin is below 1.0

**Comment by Rugvip** [comment]
@awanlin Sorry bit all over the place here, but it does indeed look like it's **not** a breaking change and so it would be a **patch** bump while we're on 0.x.

Btw, don't forget to add a changeset for the `-common` package too!

**Review by marleypowell — APPROVED** [review]
Changes look good 🙂

File: plugins/azure-devops-backend/api-report.md
Comment (freben) [inline plugins/azure-devops-backend/api-report.md:44]: No need to change this right now, but as a general hint: We have found it to be a useful pattern, especially in broad APIs like this, to pass in a single object as a parameter instead of many individual parameters. Example:

```ts
  getBuildRuns(
    options: {
      projectName: string,
      top: number,
      repoName?: string,
      definitionName?: string,
    }
  ): Promise<BuildRun[]>;
```

The outcome is that it's way easier to evolve (adding more parameters in the options structure, making things optional etc) without completely breaking the interface. You sidestep annoying pitfalls with the many-argument situation, like the impossibility of adding more arguments that are mandatory, _after_ non-mandatory ones. You also make the call sites look a lot prettier because of the named explicit object keys, so they aren't all like `.callTheThing('foo', true, true, 5, 'bar')` without any clarity to the reader what they individually mean.

I guess the azure API client itself serves as a good example, with its endless lists of `undefined` arguments, why this approach can be nice. :)

Again, just a hint for the future, feel free to ignore :)

**Review by freben — APPROVED** [review]
Took the liberty of tweaking the changeset as requested - approved, let's ship when it goes green!

**Comment by awanlin** [comment]
Thanks everyone, sorry I've been slow on this :(

**Comment by benjdlambert** [comment]
This got stuck on the 16 build waiting for an agent. Trying to get it to run now.

**Comment by benjdlambert** [comment]
Ah screw it, let's assume that 16.x passed this once rather than rebuilding the entire thing again.

**Pull request merged**
